packages/augur-core/source/contracts/BaseSimpleDex.sol

```diff
@@ -0,0 +1,182 @@
+pragma solidity 0.5.15;
+
+import "ROOT/IAugur.sol";
+import "ROOT/ISimpleDex.sol";
+import "ROOT/CashSender.sol";
+import "ROOT/libraries/token/IERC20.sol";
+import "ROOT/libraries/token/VariableSupplyToken.sol";
+import "ROOT/libraries/Initializable.sol";
+import "ROOT/libraries/ReentrancyGuard.sol";
+import "ROOT/libraries/math/SafeMathUint256.sol";
+
+
+contract BaseSimpleDex is Initializable, ReentrancyGuard, VariableSupplyToken, CashSender {
+    using SafeMathUint256  for uint256;
+
+    string constant public name = "Simple Decentralized Exchange";
+    string constant public symbol = "SDE";
+
+    IAugur public augur;
+    address public token;
+    IERC20 public cash;
+
+    uint256 public tokenReserve;
+    uint256 public cashReserve;
+    uint256 public blockNumberLast;
+    uint256 public tokenPriceCumulativeLast;
+
+    event Mint(address indexed sender, uint256 tokenAmount, uint256 cashAmount);
+    event Burn(address indexed sender, uint256 tokenAmount, uint256 cashAmount, address indexed to);
+    event Swap(address indexed sender, address indexed tokenIn, uint256 amountIn, uint256 amountOut, address indexed to);
+    event Sync(uint256 tokenReserve, uint256 cashReserve);
+
+    function transferToken(address _to, uint256 _value) private;
+
+    function getTokenBalance() public returns (uint256);
+
+    function transferCash(address _to, uint _value) private {
+        cashTransfer(_to, _value);
+    }
+
+    function getCashBalance() public view returns (uint256) {
+        return cash.balanceOf(address(this));
+    }
+
+    function initializeInternal(address _augurAddress, address _token) internal {
+        endInitialization();
+        IAugur _augur = IAugur(_augurAddress);
+        augur = _augur;
+        cash = IERC20(_augur.lookup("Cash"));
+        require(cash != IERC20(0));
+        token = _token;
+
+        initializeCashSender(_augur.lookup("DaiVat"), address(cash));
+    }
+
+    function update(uint256 _tokenBalance, uint256 _cashBalance) private {
+        uint256 _blockNumber = block.number;
+        uint256 _blocksElapsed = _blockNumber - blockNumberLast;
+        uint256 _cashReserve = cashReserve;
+        uint256 _tokenReserve = tokenReserve;
+        if (_tokenReserve != 0 && _cashReserve != 0 && _blocksElapsed > 0) {
+            // cannot reasonably overflow unless the supply of Cash, REP, or ETH became several OOM larger
+            tokenPriceCumulativeLast += _cashReserve.mul(10**18).mul(_blocksElapsed).div(_tokenReserve);
+        }
+        tokenReserve = _tokenBalance;
+        cashReserve = _cashBalance;
+        blockNumberLast = _blockNumber;
+        emit Sync(_tokenBalance, _cashBalance);
+    }
+
+    function publicMint(address _to) external nonReentrant returns (uint256 _liquidity) {
```

Comment:
The way Uniswap v2 is implemented, there are two contracts.  One is the exchange contract which offers only low-level access and is basically designed for contract interaction, not end-user interaction.  The other is a helper contract that makes the API actually usable by end users.  It doesn't appear that `publicMint` or `publicBurn` are called anywhere else in Augur at the moment, and without that wrapper contract it makes the usability of this contract really bad.

I recommend either writing your own wrapper or cribbing the Uniswap wrapper.

Reply:
As an example of one of the many usability issues, you can only burn _all_ of your tokens with `publicBurn`.  If you use a wrapper, then presumably you would transfer your tokens to the wrapper first, and then the wrapper would burn everything it had.

Reply:
Same general problem with `sellToken` and `buyToken`.

Reply:
sellToken and buyToken have convenience functions `autoSellToken` and `autoBuyToken` if thats what you mean

Reply:
right, autoBuy|Sell does the right thing, but there needs to be an auto-version of providing and removing liquidity?

Reply:
I'll consider adding those in another PR

Reply:
After talking it out with @epheph will go ahead and add these now